src/data_structures/fmindex.rs

```diff
@@ -168,18 +183,46 @@ impl FMDIndex {
     ///   less memory usage, but worse performance)
     pub fn new(bwt: BWT, k: usize) -> Self {
         let mut alphabet = dna::n_alphabet();
-        alphabet.insert(b'$');
-        assert!(
-            alphabet.is_word(&bwt),
-            "Expecting BWT over the DNA alphabet (including N) with the sentinel $."
-        );
+        alphabet.insert(SENT_CHAR);
+        assert!(alphabet.is_word(&bwt),
+                "Expecting BWT over the DNA alphabet (including N) with the sentinel $.");
 
         FMDIndex {
             fmindex: FMIndex::new(bwt, k, &alphabet),
-            revcomp: dna::RevComp::new()
+            revcomp: dna::RevComp::new(),
         }
     }
 
+    /// Find longest non-encompassing prefix match
+    pub fn longest_prefix_match(&self, pattern: &[u8], min_overlap: usize) -> Option<BiInterval> {
+        let exact_match_interval = self.fmindex.backward_search(pattern.iter());
```

Comment:
Hi Taylor,
those are very valid questions.
The rationale was that an FMIndex is a know DS, and therefore should be
a single object that owns all its parts (bwt, occ, less). This way, it
can be easily passed to other structs etc. Of course I did not have the
time so far to implement all possible algorithms on top of the FMIndex.
When sb wants to do that, he has still all possibilities on the client
side, since occ, less and bwt of the FMIndex are not hidden away but
rather available via getters (however I just noticed that I
accidentally missed the pub before occ and less!!!).
Coming to FMDIndex, it is really an extension of FMIndex that only
works on a special case, namely a text with forward and reverse
complement concatenated. Further, it is only defined on the DNA
alphabet, whereas the FMIndex is defined for arbitrary alphabets.
Hence, smems cannot be a member of FMIndex. Again, with FMDIndex, the
underlying FMIndex is accessiable via a getter, so extension is
possible without modifying the struct itself.
But of course, I would like all standard/textbook applications of FM
and FMDIndex to be in Rust-Bio. I just don't have the resources to do
that myself unless I need a particular algorithm for my own work.
However, of course I would be interested in alternative API solutions
if you have one. For example, one could think of a common trait for the
two structs. This would allow the user to conveniently add other
methods via custom traits, similar to the itertools package.
Best,
Johannes
On Tue, 2016-03-01 at 13:14 -0800, Taylor C. wrote:
> In src/data_structures/fmindex.rs:
> >          }
> >      }
> >  
> > +    /// Find longest non-encompassing prefix match
> > +    pub fn longest_prefix_match(&self, pattern: &[u8],
> min_overlap: usize) -> Option {
> > +        let exact_match_interval =
> self.fmindex.backward_search(pattern.iter());
> Additionally, what was your design intention with splitting up
> FMIndex and FMDIndex the way you did? I agree that the function I
> provided still makes sense in the context of an FMIndex, but if you
> create an FMDIndex (i.e. an FMIndex with DNA of the pattern F$R$, in
> my mind), you should still be able to perform all of the normal
> FMIndex functionality, right?
> —
> Reply to this email directly or view it on GitHub.